In [1]:
import numpy

import pandas as pd

def matrix_factorization(R, P, Q, K, steps=5000, alpha=0.0002, beta=0.02):
    Q = Q.T
    for step in range(steps):
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    eij = R[i][j] - numpy.dot(P[i,:],Q[:,j])
                    for k in range(K):
                        P[i][k] = P[i][k] + alpha * (2 * eij * Q[k][j] - beta * P[i][k])
                        Q[k][j] = Q[k][j] + alpha * (2 * eij * P[i][k] - beta * Q[k][j])
        eR = numpy.dot(P,Q)
        e = 0
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    e = e + pow(R[i][j] - numpy.dot(P[i,:],Q[:,j]), 2)
                    for k in range(K):
                        e = e + (beta/2) * (pow(P[i][k],2) + pow(Q[k][j],2))
        if e < 0.001:
            break
    return P, Q.T




In [2]:
R = [
     [5,3,0,1],
     [4,0,0,1],
     [1,1,0,5],
     [1,0,0,4],
     [0,1,5,4],
    ]


R = numpy.array(R)

N = len(R)
M = len(R[0])
K = 2

P = numpy.random.rand(N,K)
Q = numpy.random.rand(M,K)

nP, nQ = matrix_factorization(R, P, Q, K)
nR = numpy.dot(nP, nQ.T)

nR

array([[ 5.02129199,  2.86591593,  4.83289125,  0.99785853],
       [ 3.94273357,  2.25529472,  3.95717581,  0.99656205],
       [ 1.09655947,  0.73657154,  4.67312757,  4.96445062],
       [ 0.95141045,  0.6312811 ,  3.79992198,  3.97323033],
       [ 2.22692474,  1.35482141,  4.8817389 ,  4.03536193]])

In [3]:
ratings = pd.read_csv("./data/smallData_u.data", sep='\t', lineterminator='\n')
ratings

user_id  item_id  rating  timestamp
0         196      242       3  881250949
1         186      302       3  891717742
2          22      377       1  878887116
3         244       51       2  880606923
4         166      346       1  886397596
5         298      474       4  884182806
6         115      265       2  881171488
7         253      465       5  891628467
8         305      451       3  886324817
9           6       86       3  883603013
10         62      257       2  879372434
11        286     1014       5  879781125
12        200      222       5  876042340
13        210       40       3  891035994
14        224       29       3  888104457
15        303      785       3  879485318
16        122      387       5  879270459
17        194      274       2  879539794
18        291     1042       4  874834944
19        234     1184       2  892079237
20        119      392       4  886176814
21        167      486       4  892738452
22        299      144       4  877881320
23        291      118       2  874833878
24        308        1       4  887736532
25         95      546       2  879196566
26         38       95       5  892430094
27        102      768       2  883748450
28         63      277       4  875747401
29        160      234       5  876861185
...       ...      ...     ...        ...
9970      318      419       5  884495890
9971        1      150       5  876892196
9972      380      177       3  885479082
9973       23      549       3  874788290
9974       65      173       3  879217851
9975       56      435       3  892676429
9976      210      403       4  887736322
9977      147      340       4  885593965
9978      339       28       4  891032542
9979       12      753       5  879960679
9980        2      308       3  888979945
9981      264      283       5  886122952
9982      345      315       5  884900653
9983      193      111       1  889126375
9984      301        7       4  882074236
9985      373       68       5  877106741
9986       94       38       2  891722482
9987      357      833       4  878952341
9988       26      237       3  891351590
9989      316      197       4  880854227
9990      293      164       4  888906598
9991      217      226       1  889069878
9992      145      413       3  877343280
9993      371      177       4  877487135
9994      286      477       3  876521773
9995       74      150       3  888333458
9996      178      195       4  882826944
9997      321      527       3  879439763
9998      337      742       5  875184353
9999       90      190       5  891383687

[10000 rows x 4 columns]

In [4]:
uniqueUserIdList = ratings.user_id.unique()
uniqueUserIdList

array([196, 186,  22, 244, 166, 298, 115, 253, 305,   6,  62, 286, 200,
       210, 224, 303, 122, 194, 291, 234, 119, 167, 299, 308,  95,  38,
       102,  63, 160,  50, 301, 225, 290,  97, 157, 181, 278, 276,   7,
        10, 284, 201, 287, 246, 242, 249,  99, 178, 251,  81, 260,  25,
        59,  72,  87,  42, 292,  20,  13, 138,  60,  57, 223, 189, 243,
        92, 241, 254, 293, 127, 222, 267,  11,   8, 162, 279, 145,  28,
       135,  32,  90, 216, 250, 271, 265, 198, 168, 110,  58, 237,  94,
       128,  44, 264,  41,  82, 262, 174,  43,  84, 269, 259,  85, 213,
       121,  49, 155,  68, 172,  19, 268,   5,  80,  66,  18,  26, 130,
       256,   1,  56,  15, 207, 232,  52, 161, 148, 125,  83, 272, 151,
        54,  16,  91, 294, 229,  36,  70,  14, 295, 233, 214, 192, 100,
       307, 297, 193, 113, 275, 219, 218, 123, 158, 302,  23, 296,  33,
       154,  77, 270, 187, 170, 101, 184, 112, 133, 215,  69, 104, 240,
       144, 191,  61, 142, 177, 203,  21, 197, 134, 180, 236, 26

In [5]:
uniqueUserCount = len(ratings.user_id.unique())
uniqueUserCount

385

#### Our A matrix has 385 rows. Column number would be equal to number of movies

In [6]:
movies = pd.read_csv("./data/u.item", sep='|', lineterminator='\n',encoding='cp1252')
movies

movie_id                                        movie_title  \
0            1                                   Toy Story (1995)   
1            2                                   GoldenEye (1995)   
2            3                                  Four Rooms (1995)   
3            4                                  Get Shorty (1995)   
4            5                                     Copycat (1995)   
5            6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...   
6            7                              Twelve Monkeys (1995)   
7            8                                        Babe (1995)   
8            9                            Dead Man Walking (1995)   
9           10                                 Richard III (1995)   
10          11                               Seven (Se7en) (1995)   
11          12                         Usual Suspects, The (1995)   
12          13                            Mighty Aphrodite (1995)   
13          14                                 Postino, Il (1994)   
14          15                          Mr. Holland's Opus (1995)   
15          16                 French Twist (Gazon maudit) (1995)   
16          17                         From Dusk Till Dawn (1996)   
17          18                          White Balloon, The (1995)   
18          19                              Antonia's Line (1995)   
19          20                          Angels and Insects (1995)   
20          21                      Muppet Treasure Island (1996)   
21          22                                  Braveheart (1995)   
22          23                                 Taxi Driver (1976)   
23          24                         Rumble in the Bronx (1995)   
24          25                               Birdcage, The (1996)   
25          26                      Brothers McMullen, The (1995)   
26          27                                    Bad Boys (1995)   
27          28                                   Apollo 13 (1995)   
28          29                              Batman Forever (1995)   
29          30                               Belle de jour (1967)   
...        ...                                                ...   
1652      1653  Entertaining Angels: The Dorothy Day Story (1996)   
1653      1654                       Chairman of the Board (1998)   
1654      1655                                  Favor, The (1994)   
1655      1656                                 Little City (1998)   
1656      1657                                      Target (1995)   
1657      1658                      Substance of Fire, The (1996)   
1658      1659                    Getting Away With Murder (1996)   
1659      1660                                 Small Faces (1995)   
1660      1661                                New Age, The (1994)   
1661      1662                                 Rough Magic (1995)   
1662      1663                            Nothing Personal (1995)   
1663      1664                     8 Heads in a Duffel Bag (1997)   
1664      1665                           Brother's Kiss, A (1997)   
1665      1666                                        Ripe (1996)   
1666      1667                              Next Step, The (1995)   
1667      1668                          Wedding Bell Blues (1996)   
1668      1669                           MURDER and murder (1996)   
1669      1670                                     Tainted (1998)   
1670      1671                          Further Gesture, A (1996)   
1671      1672                                        Kika (1993)   
1672      1673                                      Mirage (1995)   
1673      1674                                  Mamma Roma (1962)   
1674      1675                              Sunchaser, The (1996)   
1675      1676                            War at Home, The (1996)   
1676      1677                               Sweet Nothing (1995)   
1677      1678                                  Mat' i syn (1997)   
1678

In [7]:
movies.index

RangeIndex(start=0, stop=1682, step=1)

#### We have 1682 movies. This will be our column count in the A matrix.

In [ ]:
# Constructing A matrix
A = []
# movie_data_limit = 1682
movie_data_limit = 30  # for fast calculation
for i in range(len(uniqueUserIdList)):
    userId = uniqueUserIdList[i]
    tmpDF = ratings.loc[ratings['user_id'] == userId]
    row = []
    for j in range(movie_data_limit):
        movieId = i+1 # movie Id starts with 1
        movieRatingDF = tmpDF.loc[tmpDF['item_id'] == movieId]
        if movieRatingDF.empty:
            row.append(0)
        else:
            val = movieRatingDF.iloc[0][2]
            row.append(val)
        A.append(row)
print(A)
    



[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [ ]:
R = numpy.array(A)

N = len(R)
M = len(R[0])
K = 2

P = numpy.random.rand(N,K)
Q = numpy.random.rand(M,K)

nP, nQ = matrix_factorization(R, P, Q, K)
nR = numpy.dot(nP, nQ.T)

nR